File: python/Q25_reverse_nodes_in_k_group.py
#
# @lc app=leetcode id=25 lang=python3
#
# [25] Reverse Nodes in k-Group
#


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val: int = val
        self.next: ListNode | None = next


# @lc code=start
class Solution:
    def longer_than_k(self, head: ListNode, k: int) -> bool:
        for _ in range(k):
            if not head.next:
                return False

            head = head.next

        return True

    def reverse(self, head: ListNode, k: int) -> ListNode:
        previous: ListNode = head
        assert head.next is not None
        current: ListNode | None = head.next
        reversed_head: ListNode | None = head.next

        for _ in range(k):
            assert current is not None
            tmp: ListNode | None = current.next
            current.next = previous
            previous = current
            current = tmp

        head.next = previous
        reversed_head.next = current
        return reversed_head

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy: ListNode = ListNode(next=head)
        current: ListNode = dummy

        while self.longer_than_k(current, k):
            current = self.reverse(current, k)

        return dummy.next

    def reverseKGroupRecursive(
        self, head: Optional[ListNode], k: int
    ) -> Optional[ListNode]:
        count: int = 0
        counting_node: ListNode | None = head

        while counting_node:
            count += 1
            if count == k:
                break

            counting_node = counting_node.next

        if count < k:
            return head

        previous: ListNode | None = None
        current: ListNode | None = head

        for _ in range(k):
            assert current is not None
            tmp: ListNode | None = current.next
            current.next = previous
            previous = current
            current = tmp

        if head:
            head.next = self.reverseKGroup(current, k)

        return previous


# @lc code=end

File: rust/src/leetcode/Q25_reverse_nodes_in_k_group.rs
/*
 * @lc app=leetcode id=25 lang=rust
 *
 * [25] Reverse Nodes in k-Group
 */

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @lc code=start
impl Solution {
    pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut count: i32 = 0;
        let mut current: &Option<Box<ListNode>> = &head;
        while let Some(node) = current {
            count += 1;
            if count == k {
                break;
            }
            current = &node.next;
        }

        if count < k {
            return head;
        }

        let mut previous: Option<Box<ListNode>> = None;
        let mut current: Option<Box<ListNode>> = head;

        for _ in 0..k {
            if let Some(mut node) = current {
                let next: Option<Box<ListNode>> = node.next.take();
                node.next = previous;
                previous = Some(node);
                current = next;
            }
        }

        if let Some(ref mut tail) = previous {
            let mut tail_node: &mut Box<ListNode> = tail;
            while tail_node.next.is_some() {
                tail_node = tail_node.next.as_mut().unwrap();
            }
            tail_node.next = Self::reverse_k_group(current, k);
        }

        previous
    }
}
// @lc code=end
